tests/integration.rs

```diff
@@ -0,0 +1,433 @@
+                )
+                .https_or_http()
+                .enable_http1()
+                .enable_http2()
+                .build(),
+        )
+    }
+
+    // Provision an HTTP-01 challenge response for the given token and key authorization.
+    //
+    // The Pebble challenge test server will be configured to respond to HTTP-01 challenge
+    // requests for the provided token by returning the expected key auth.
+    async fn add_http01_response(
+        &self,
+        token: &str,
+        key_auth: &str,
+    ) -> Result<(), Box<dyn StdError>> {
+        let client = self.http_client();
+
+        let body = json!({
+            "token": token,
+            "content": key_auth,
+        })
+        .to_string();
+
+        let req = Request::builder()
+            .method(Method::POST)
+            .uri(format!("{}/add-http01", self.challenge_management_url()))
+            .header(CONTENT_TYPE, "application/json")
+            .body(Full::new(body.into()))?;
+
+        client.request(req).await?;
+
+        Ok(())
+    }
+
+    // Return a RootCertStore containing the issuer root for the Pebble CA.
+    //
+    // This is distinct from the management root CA, and is randomly generated each
+    // time that Pebble starts up. This is the issuer that signs the randomly generated
+    // intermediate certificate returned as part of ACME issued certificate chains.
+    async fn issuer_roots(&self) -> Result<RootCertStore, Box<dyn StdError>> {
+        let client = self.http_client();
+
+        let req = Request::builder()
+            .method(Method::GET)
+            .uri(format!("{}/roots/0", self.pebble_management_url()))
+            .header(CONTENT_TYPE, "application/json")
+            .body(Full::default())?;
+
+        let resp = client.request(req).await?;
+        if resp.status() != 200 {
+            return Err(format!("unexpected /roots/0 response status: {}", resp.status()).into());
+        }
+
+        let body = resp.collect().await?.aggregate();
+        let mut pem = String::new();
+        body.reader().read_to_string(&mut pem)?;
+        let root = CertificateDer::from_pem_slice(pem.as_bytes())?;
+
+        let mut roots = RootCertStore::empty();
+        roots.add_parsable_certificates(vec![root]);
+        assert_eq!(roots.len(), 1);
+        Ok(roots)
+    }
+
+    fn challenge_management_url(&self) -> &str {
+        "http://127.0.0.1:8055" // Default.
+    }
+
+    fn pebble_management_url(&self) -> String {
+        format!("https://{}", &self.config.pebble.management_listen_address)
+    }
+
+    fn directory_url(&self) -> String {
+        format!("https://{}/dir", &self.config.pebble.listen_address)
+    }
+}
+
+#[derive(Clone, Serialize)]
+struct Config {
+    pebble: PebbleConfig,
+}
+
+#[derive(Clone, Serialize)]
+struct PebbleConfig {
+    #[serde(rename = "listenAddress")]
+    listen_address: String,
+    #[serde(rename = "managementListenAddress")]
+    management_listen_address: String,
```

Comment:
Could we use `SocketAddr` for these?

Reply:
I don't think we can `const` construct a `SocketAddr` so I left this as a `&'static str`. It also feels like it's of less value in this case since we're mostly using this as a string to construct a URL, and not binding the address ourselves outside of the loop that polls for the port to be ready.